### PR TITLE
Minor improvements to least_squares' docstring.

### PR DESCRIPTION
**Pull request by anntzer**
Fixing a few typos and wording.

Note that English isn't my native language either so you may want to run this through someone else.

More importantly, I think the `scaling` parameter is a bit confusing (I left that part of the docs untouched for now).  At least I personally tend to think of such a parameter as providing the "scale" over which the parameters vary, ie. it should have the _same_ physical unit as the parameter itself.  But if I read the docs correctly, it seems that it is the inverse of that, with _inverse_ units?  Not sure what to do about this tough.

@nmayorov @ev-br

### CODE REVIEW

File: scipy/optimize/_lsq/least_squares.py [hunk 1]
```diff
@@ -234,45 +234,40 @@ def least_squares(
         kwargs={}):
     """Solve a nonlinear least-squares problem with bounds on the variables.
 
-    Let f(x) be an m-dimensional function of n variables and rho(s) be a
-    scalar function, `least_squares` finds a local minimum of::
+    Given f(x), an m-dimensional function of n variables (the residuals),
+    and rho(s), a scalar function (the loss function), `least_squares`
+    finds a local minimum of the cost function F(x)::
```
Comment (nmayorov) [inline scipy/optimize/_lsq/least_squares.py]: I like this new description a lot better. 

You decided to drop Jacobian definition here. It looks logical, but on the other hand Jacobian is probably the most important concept for the whole algorithm (my suggestion how to resolve it below).

File: scipy/optimize/_lsq/least_squares.py [hunk 2]
```diff
@@ -234,45 +234,40 @@ def least_squares(
...
-        function is ndarray of shape (n,) (never a scalar, even for n=1).
+        respect to its first argument. The argument ``x`` passed to this
+        function is an ndarray of shape (n,) (never a scalar, even for n=1).
         It must return a 1-d array_like of shape (m,) or a scalar.
     x0 : array_like with shape (n,) or float
         Initial guess on independent variables. If float, it will be treated
         as a 1-d array with one element.
     jac : {'2-point', '3-point', 'cs', callable}, optional
-        Method of computing the Jacobian matrix. The keywords select the
+        Method of computing the Jacobian matrix. The keywords select a
```
Comment (nmayorov) [inline scipy/optimize/_lsq/least_squares.py]: I think it's better to give a definition of Jacobian here, not on the very last sentence.

File: scipy/optimize/_lsq/least_squares.py [hunk 3]
```diff
@@ -446,7 +441,7 @@ def least_squares(
             * -1 : a lower bound is active.
             *  1 : an upper bound is active.
 
-        Might be somewhat arbitrary for 'trf' method as it does strictly
+        Might be somewhat arbitrary for 'trf' method as it only uses strictly
```
Comment (nmayorov) [inline scipy/optimize/_lsq/least_squares.py]: I searched "Numerical Optimization" book for the word "iterates", it is used mostly as or "generates a sequence of iterates"

So I suggest "as it generates a sequence of strictly feasible iterates" ("uses" is also good, "does" is not correct).

**Comment by nmayorov** [comment]
@anntzer your concern about "scaling" makes sense, I remember having some back and forths deciding how to define it. 

As I remember, I finally decided to define it according to `diag` parameter of `leastsq` (and MINPACK originally): "N positive entries that serve as a scale factors for the variables."

I guess it's not too late to turn it other way. I see that it can have benefits for consistency with `loss_scale`. @ev-br what do you think?

**Comment by nmayorov** [comment]
On the other hand, `loss_scale` and `scaling` are quite different concepts (one applies to residuals, another to variables), so maybe it's even better to define them differently.

**Comment by anntzer** [comment]
As a physicist, I think that arguments should be expressed in the same units as often as possible, even if they have different roles.

**Comment by ev-br** [comment]
Maybe Chuck would be able to weigh in on the matters of proper English? @charris

**Comment by anntzer** [comment]
I assume the milestone means it's too late too change the API to have the scaling in the same dimension as the parameters?

Regarding the scaling, I noticed that the meaning of setting it to "jac" was also a bit unclear to me from the text.  In my (admittedly limited) understanding of the problem, you basically want to solve the problem using scaled parameters (x_i -> x_i / s_i) so that dF/dx_i = 1 for all i, right?  In my old, naïve way of doing it, if I needed this, I would solve the problem once, and use the final jacobian to reparametrize the problem - which I could do here by passing the norm of the jacobian of the first solution to "scaling".

But what happens when you actually set `scaling="jac"`?  Do you recompute the scaling at each iteration?

**Comment by nmayorov** [comment]
The "jac" scaling method is taken from "The Levenberg-Marquardt algorithm: Implementation and Theory", sec. 6.

The author claims that good options for scaling are:

a) s_i = || d f(x_0) / d x_i || (the Euclidean norm of Jacobian columns at the starting point)
b) Or alternatively s_i is updated on each iteration, using the formula a), but if the new scaling value is less --- it is not updated.

The approach b) was chosen in MINPACK, I decided to follow it too. 

I suppose that your method of recomputing can also be useful. Generally there is no guarantee that any method will give you a "right" solution.

I also confirm that your understanding of the `scaling` as `x_i -> x_i / s_i` is correct. Do you still find it inconvenient? @ev-br your opinion would be helpful, I tend to agree with Antony that changing to `scale` will make it easier to understand.

**Comment by ev-br** [comment]
> I assume the milestone means it's too late too change the API to have the scaling in the same dimension as the parameters?

Not necessarily. I won't hold the release for this, but if it's ready in time, I see no problem backporting it.

File: scipy/optimize/_lsq/least_squares.py (hunk 1 above)
Comment (ev-br) [inline scipy/optimize/_lsq/least_squares.py]: Nitpick: I would tend to read `m-dimensional function of n variables (the residuals)` as if these n variables are themselves residuals.

**Comment by ev-br** [comment]
Re scaling meaning scale or inverse scale: a physisist in me does not have a prefetence; an occasional user in me would anyway forget which one it is between uses, and would look it up in the docs --- please document it.

**Comment by nmayorov** [comment]
@ev-br it is documented.

**Comment by anntzer** [comment]
OK, I applied some changes accordingly and moved the definition of "scaling" (as inverse scale) to the beginning of the section.  I think this may be clearer for the user.

**Comment by nmayorov** [comment]
About `scaling` vs `scale`. I wonder what would be the best way for the community. I don't know... How important is to be consistent with `leastsq'` parameter `diag`? Probably not very important.

@anntzer could you write here the description of `scale` how you see it? Maybe it will be easier to choose then.

**Comment by anntzer** [comment]
Something like...

```
Relative characteristic scale of each variable. Improved convergence may
be achieved by rescaling each variable so that a step of a given size on
any one of them has a similar effect on the cost function. Alternatively,
`scaling` can be thought of as the inverse of the diagonal elements of a
matrix that determines the shape of a trust region <actually "determines"
is rather vague here>. Default is 1.0, i.e., no scaling: a scalar value has
no effect (except on possible numerical issues and by changing termination
criteria). If set to "jac", the scaling is iteratively updated using the
inverse norms of the columns of the Jacobian matrix (as described in
[JJMore]_).
```

**Comment by nmayorov** [comment]
Better to name it `scale` then? (@ev-br is it possible?)

About a trust-region: `An alternative <or rather complementary> view is that the size of a trust-region along j-th dimension is proportional to scale[j]`. Basically, a trust region is stretched proportional to `scale`. (Again, you sure can compose a better wording).

Clarification about "the shape of a trust-region": If `D` is a diagonal matrix containing current `scaling` elements, then a trust region is defined as `||D s|| < Delta`, so `D` contains values inversely proportional  to axis lengths of a trust-region ellipsoid.

**Comment by anntzer** [comment]
I like your wording about the trust region much better.

**Comment by anntzer** [comment]
Actually, I found another point that is unclear: you define `opt.jac` as "Modified Jacobian matrix at the solution, in the sense that J^T J is a Gauss-Newton approximation of the Hessian of the cost function."  What is the modification?  AFAIK the Gauss-Newton approximation is exactly to approximate H as J^T J, right?

**Comment by ev-br** [comment]
> Better to name it scale then? (@ev-br is it possible?)

Until it's released, yes, easily. https://github.com/scipy/scipy/pull/5696#issuecomment-170337424

**Comment by nmayorov** [comment]
@anntzer, about `res.jac`. 

Note, that the cost function is the sum of loss functions applied to squares of the residuals. Thus, "the Gauss-Newton part" of Hessian of F is not J^T J, it also depends on the loss function. 

The loss function handling is implemented by modifying Jacobian on each iteration, such that J'^T J' (note primes) matches with Hessian part of. So basically `res.jac` contains this inner modified Jacobian. Sure it is the standard Jacobian when `loss="linear"`.

You can see some formulas here http://ceres-solver.org/nnls_modeling.html#theory

**Comment by nmayorov** [comment]
OK, I suggest to change the definition of `scaling` to its inverse. I propose two possible renames:

1. `scaling -> scale`
2. `scaling -> x_scale`, `loss_scale -> f_scale`

Opinions?

When doing re-definition it is better (in my opinion) too make it consistent for all code related to `least_squares`, i. e. not just pass `1 / scale` to implementation methods. It will keep code more maintainable and clean. 

@anntzer you wouldn't mind to do this maintain work (can be a bit tedious)? You can spot some bad code or bugs, will be useful.

@rgommers, sorry to bother you, could you give your opinion about this redefinition / rename?

**Comment by anntzer** [comment]
I think `x_scale` is good but will wait until one of @ev-br or @rgommers give their opinion before starting the search-replace fest.
(Apparently the code uses `scaling` and `scale` more or less with the same meaning, e.g. you wrote `scale, scale_inv = scaling, 1 / scaling`.)

**Comment by ev-br** [comment]
The doc changes LGTM.

Regarding the `scale` changes: I'm fine either way, but the time can be a bit tight. I'd like to cut the release once gh-5697 is done.

**Comment by anntzer** [comment]
I think I'll give up on it; I don't have much time this week-end and documenting it properly is probably more than enough (and anyways I already gave up on changing the bounds API).

**Comment by nmayorov** [comment]
@ev-br I think you should merge this now. And I'll send another PR with redefinition / rename till tomorrow's evening.

**Comment by ev-br** [comment]
Sounds like a plan :-). Merging, thank you Anthony, Nikolay

**Comment by ev-br** [comment]
OK, backported in gh-5726

**Comment by anntzer** [comment]
@ev-br Any news regarding the date of the release of 0.17?  Turns out there's some code I wrote that relies on `least_squares` and that I'd like to share with some colleagues :)
Obviously there's no hurry given that I participated in creating more work, I'd just like a rough estimate.  Thanks!

**Comment by rgommers** [comment]
looks like @ev-br did hurry, 0.17.0 is out:)

**Comment by anntzer** [comment]
I saw that, thanks!

**Pull request merged**